mongodb/genny
# Modernize genny python and paths

This is bigger than I was hoping, but I think it's straightforward enough. Getting this out there for some early feedback.

Related:

- https://mongodbcr.appspot.com/749470020/
- https://github.com/10gen/dsi/pull/826 
- https://github.com/mongodb/genny/pull/412

A few things going on here:

- all genny invocations go through `run-genny` which is basically a straight lift of `run-dsi`.
- The existing `lamp` commands have been folded into the python that `run-genny` invokes. That now uses click for arg parsing.
- All of the shell scripts have been ported to python.
- **Note**: A lot of the python diff is just moving things around. The only new python is `cli.py`, `resmoke.py`, and `dry_run.py`, 
- The default `Metrics.Path` is now `build/CedarMetrics`

Simply changing the metrics path to `./build/CedarMetrics` turned rather unpleasant because we run genny either from the genny repo root or from pwd having repo root of `./src/genny`. With everything going through `run-genny` and being in python we can now easily compute paths and know exactly where things are.

A few minor things:

- Moved logging to structlog
- Moved testing to pytest
- The evergreen config no longer uploads to S3 (it was never used)
- The evergreen config now uses `run-genny` via `subprocess.exec`
- The mongo repo (used by resmoke tests) is now an evergreen module

Still remaining before merge:

- Ensure this works in a sane way for running from either the genny repo root or from the new evergreen conventional location which has `./src/genny` as the repo root.
- Update paths in DSI (mostly done as I did it in a branch in parallel with these changes)
- Update docs
- Add pytests for some of the nontrivial logic added (although there is very little real logic)
- Of course I'll ensure CI tests pass - going to be a bit annoying with broken compile but I'll manage

If this PR feels too big, I can split it into:

- kill s3 in evergreen
- mongo is a module
- run everything through run-genny
- moving to structlog and pytest
- moving shell to python

But that would be a fair amount of coordination with the DSI repo and doesn't seem worth it. Please push back if you disagree.
